File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        ListNode* dummy = new ListNode(0, head);
        ListNode* pre = dummy;
        ListNode* next_node = nullptr;
        ListNode*  cur = nullptr;

        for(auto i = 0; i <left-1; i++){
            pre = pre->next;
        }
        cur = pre->next;
        
        for(auto i=left; i < right; i++){
            next_node = cur->next;
            cur->next = next_node->next;
            next_node->next = pre->next;
            pre->next = next_node;
        }
        return dummy->next;
    }
};

File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if right == left:
            return head

        slow = head
        dummy = ListNode(next=slow)
        prev = dummy

        for _ in range(left - 1):
            prev = prev.next
        
        cur = prev.next
        prev_node = None

        for _ in range(left, right + 1):
            next_ptr = cur.next
            cur.next = prev_node
            prev_node = cur
            cur = next_ptr
        
        prev.next.next = cur
        prev.next = prev_node

        return dummy.next



       

        
        
        